lib/token-store.js

```diff
@@ -8,196 +8,1110 @@
 // Implementation
 //------------------------------------------------------------------------------
 
-module.exports = function(tokens) {
-    const api = {},
-        starts = Object.create(null),
-        ends = Object.create(null),
-        length = tokens.length;
+const W_INIT = -3;
+const W_TOKENS = 1;
+const W_COMMENTS = 2;
+const PUBLIC_METHODS = Object.freeze([
+    "getTokenByRangeStart",
+
+    "getFirstToken",
+    "getLastToken",
+    "getTokenBefore",
+    "getTokenAfter",
+    "getFirstTokenBetween",
+    "getLastTokenBetween",
+
+    "getFirstTokens",
+    "getLastTokens",
+    "getTokensBefore",
+    "getTokensAfter",
+    "getFirstTokensBetween",
+    "getLastTokensBetween",
+
+    "getTokens",
+    "getTokensBetween",
+
+    "getTokenOrCommentBefore",
+    "getTokenOrCommentAfter"
+]);
+
+/**
+ * Binary-searches the index of the first token which is after the given location.
+ * If it was not found, this returns `tokens.length`.
+ *
+ * @param {(Token|Comment)[]} tokens - It searches the token in this list.
+ * @param {number} location - The location to search.
+ * @returns {number} The found index or `tokens.length`.
+ * @private
+ */
+function search(tokens, location) {
+    let left = 0;
+    let right = tokens.length - 1 | 0;
+
+    while (left <= right) {
+        const middle = (left + right) / 2 | 0;
+        const value = tokens[middle].range[0] | 0;
+
+        if (value < location) {
+            left = middle + 1 | 0;
+        } else if (value > location) {
+            right = middle - 1 | 0;
+        } else {
+            return middle;
+        }
+    }
+
+    return (left > right) ? left : right;
+}
+
+/**
+ * The abstract class about cursors which iterate tokens.
+ *
+ * This class has 3 abstract methods.
+ *
+ * - `current: Token | Comment` ... The current token.
+ * - `moveNext(): boolean` ... Moves this cursor to the next token. If the next token didn't exist, it returns `false`.
+ * - `movePrev(): boolean` ... Moves this cursor to the previous token. If the previous token didn't exist, it returns `false`.
+ *
+ * There are the following known sub classes.
+ *
+ * - BaseCursor
+ *     - TokenCursor ......... The cursor which iterates tokens only.
+ *     - TokenCommentCursor .. The cursor which iterates tokens and comments.
+ * - DecorativeCursor
+ *     - FilterCursor ........ The cursor which ignores the specified tokens.
+ *     - SkipCursor .......... The cursor which ignores the first few tokens.
+ *     - LimitCursor ......... The cursor which limits the count of tokens.
+ *
+ * @private
+ */
+class Cursor {
+
+    /**
+     * Initializes this cursor.
+     */
+    constructor() {
+        this.current = null;
```

Comment:
I think it might be more readable if `this.current` was called `this.currentToken` (or maybe `this.currentValue`) instead. At first, I couldn't tell whether `this.current` was referring to an index.